package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,690 @@
+ctypedef real matrix[DIM][DIM]
+
+
+# Useful Functions
+cdef real rvec_norm2(const rvec a) nogil:
+    return a[XX]*a[XX]+a[YY]*a[YY]+a[ZZ]*a[ZZ]
+
+cdef void rvec_clear(rvec a) nogil:
+    a[XX]=0.0
+    a[YY]=0.0
+    a[ZZ]=0.0
+
+###############################
+# Utility class to handle PBC #
+###############################
+cdef struct cPBCBox_t:
+    matrix     box
+    rvec       fbox_diag
+    rvec       hbox_diag
+    rvec       mhbox_diag
+    real       max_cutoff2
+
+
+# Class to handle PBC calculations
+cdef class PBCBox(object):
+    cdef cPBCBox_t c_pbcbox
+    cdef rvec center
+    cdef rvec bbox_center
+    cdef bint is_triclinic
+
+    def __init__(self, real[:,::1] box):
+        self.update(box)
+
+    cdef void fast_update(self, real[:,::1] box) nogil:
+        cdef ns_int i, j, k, d, jc, kc, shift
+        cdef real d2old, d2new, d2new_c
+        cdef rvec trial, pos
+        cdef ns_int ii, jj ,kk
+        cdef ns_int *order = [0, -1, 1, -2, 2]
+        cdef bint use
+        cdef real min_hv2, min_ss, tmp
+
+        rvec_clear(self.center)
+        # Update matrix
+        self.is_triclinic = False
+        for i in range(DIM):
+            for j in range(DIM):
+                self.c_pbcbox.box[i][j] = box[i, j]
+                self.center[j] += 0.5 * box[i, j]
+
+                if i != j:
+                    if box[i, j] > EPSILON:
+                        self.is_triclinic = True
+            self.bbox_center[i] = 0.5 *  box[i, i]
+
+        # Update diagonals
+        for i in range(DIM):
+            self.c_pbcbox.fbox_diag[i]  =  box[i, i]
+            self.c_pbcbox.hbox_diag[i]  =  self.c_pbcbox.fbox_diag[i] * 0.5
+            self.c_pbcbox.mhbox_diag[i] = - self.c_pbcbox.hbox_diag[i]
+
+        # Update maximum cutoff
+
+        # Physical limitation of the cut-off
+        # by half the length of the shortest box vector.
+        min_hv2 = min(0.25 * rvec_norm2(&box[XX, XX]), 0.25 * rvec_norm2(&box[YY, XX]))
+        min_hv2 = min(min_hv2, 0.25 * rvec_norm2(&box[ZZ, XX]))
+
+        # Limitation to the smallest diagonal element due to optimizations:
+        # checking only linear combinations of single box-vectors (2 in x)
+        # in the grid search and pbc_dx is a lot faster
+        # than checking all possible combinations.
+        tmp = box[YY, YY]
+        if box[ZZ, YY] < 0:
+            tmp -= box[ZZ, YY]
+        else:
+            tmp += box[ZZ, YY]
+
+        min_ss = min(box[XX, XX], min(tmp, box[ZZ, ZZ]))
+        self.c_pbcbox.max_cutoff2 = min(min_hv2, min_ss * min_ss)
+
+    def update(self, real[:,::1] box):
+        if box.shape[0] != DIM or box.shape[1] != DIM:
+            raise ValueError("Box must be a {} x {} matrix. Got: {} x {})".format(
+                DIM, DIM, box.shape[0], box.shape[1]))
+        if (box[XX, XX] == 0) or (box[YY, YY] == 0) or (box[ZZ, ZZ] == 0):
+            raise ValueError("Box does not correspond to PBC=xyz")
+        self.fast_update(box)
+

```

Comment:
I think a functionality to handle no pbc can be implemented here. Probably something like 
```
def fast_update_nopbc(box):
    min_dim = coords.min(axis=0)
    max_dim = coords.max(axis=0)
    box[:3] = max_dim.max(axis=0) - min_dim.min(axis=0)
    box[3:] = 90. # making it an orthogonal box
```
and handle the same in search function.